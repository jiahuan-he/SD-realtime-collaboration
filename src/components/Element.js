import React from 'react'

const stock = {
    stroke: 'black',
    cursor: 'move',
    strokeWidth: "1px",
    fill: 'white'
}

const stockHighlight = {
    stroke: 'green',
    cursor: 'move',
    strokeWidth: "2px",
    fill: 'white'
}

const moveable = {
    cursor: 'move',
}

export default class Element extends React.Component {
    componentDidMount() {
        this.makeDraggable()
    }

    makeDraggable = () => {
        const id = this.props.stock.id
        let draggable = document.getElementById(id);
        let offset

        const mouseDown = (e) => {
            console.log("mouse down")
            draggable = document.getElementById(id);
            offset = getMousePosition(e);
            offset.x -= parseFloat(draggable.getAttributeNS(null, "x"));
            offset.y -= parseFloat(draggable.getAttributeNS(null, "y"));
            draggable.addEventListener('mousemove', mouseMove)
        }

        const mouseMove = (e) => {
            console.log("mouse move")
            if (!draggable) return
            e.preventDefault();
            const coord = getMousePosition(e);
            console.log(coord.x - offset.x)
            console.log(coord.y - offset.y)
            this.props.updatePosition(id, coord.x - offset.x, coord.y - offset.y)
        }

        const mouseUpOrLeave = (e) => {
            console.log("mouse up")
            draggable = null
        }

        // mouse position to svg position
        const getMousePosition = (e) => {
            const CTM = draggable.getScreenCTM();
            return {
                x: (e.clientX - CTM.e) / CTM.a,
                y: (e.clientY - CTM.f) / CTM.d
            };
        }

        draggable.addEventListener('mousedown', mouseDown)
        draggable.addEventListener('mouseup', mouseUpOrLeave)
        draggable.addEventListener('mouseleave', mouseUpOrLeave)
    }

    render() {
        const x=this.props.stock.posX
        const y=this.props.stock.posY
        const id=this.props.stock.id
        const stockValue=this.props.stock.value
        return (
            <g 
                x={x} 
                y={y} 
                id={id} 
                style={moveable}
            >
                <rect x={x} y={y} width={"5%"} height={"5%"} style={this.props.highlight?stockHighlight:stock}/>
                <foreignObject x={x} y={y} width="50" height="50">
                    <div>{id+": "+stockValue}</div>                 
                </foreignObject>    
            </g>       
        );
    }
}