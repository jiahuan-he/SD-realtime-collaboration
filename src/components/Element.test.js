import React from 'react';
import { mount } from 'enzyme'
import Element from './Element'
// Avoid Warning: render(): Rendering components directly into document.body is discouraged.
beforeAll(() => {
    const div = document.createElement('div');
    window.domNode = div;
    document.body.appendChild(div);
  })

describe('Element', () => {
    it('should render correctly in "debug" mode', () => {
        const props = {
            id: "stock0",
            stockValue: 0,
            x: 50,
            y: 50,
        }
        const element = mount(<svg>
            <Element
                key={props.id}
                stockID={props.id}
                stockValue={props.stockValue}
                x={props.x}
                y={props.y}
            debug
            ></Element></svg>, { attachTo: window.domNode });
        expect(element).toMatchSnapshot();
    });
});