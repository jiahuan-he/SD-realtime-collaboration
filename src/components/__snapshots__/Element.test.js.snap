// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element should render correctly in "debug" mode 1`] = `
<svg>
  <Element
    debug={true}
    key="stock0"
    stockID="stock0"
    stockValue={0}
    x={50}
    y={50}
  >
    <g
      id="stock0"
      style={
        Object {
          "cursor": "move",
        }
      }
      x={50}
      y={50}
    >
      <rect
        height="5%"
        style={
          Object {
            "cursor": "move",
            "fill": "white",
            "stroke": "black",
            "strokeWidth": "1px",
          }
        }
        width="5%"
        x={50}
        y={50}
      />
      <foreignObject
        height="50"
        width="50"
        x={50}
        y={50}
      >
        <div>
          stock0: 0
        </div>
      </foreignObject>
    </g>
  </Element>
</svg>
`;
